plat: rcar: enable CORE ASLR

lorc opened the pull request:
This set of patches enables various features on Renesas RCAR platform to enable core ASLR.

During testing it appeared that `CFG_VIRTUALIZATION` conflicts with `CFG_CORE_ASLR` , so for time being they are made mutually exclusive. I'm looking at this issue right now.

lorc commented:
There is a `checkpatch` warning about `else` after `return`. I'm deliberately left it, because in this case function looks symmetric.
If there are no objections, I would prefer to leave it as is.

lorc commented:
And another, slightly related topic:

`plat-rcar/main.c` have full BSD-2 license at the top. If I get it correct, we can remove it only if all GlobaLogic agree on this. Or there is no such restriction?

jenswi-linaro commented:
> And another, slightly related topic:
> 
> `plat-rcar/main.c` have full BSD-2 license at the top. If I get it correct, we can remove it only if all GlobaLogic agree on this. Or there is no such restriction?

If it matches word by word with the official BSD-2 license then it should be OK to use `SPDX-License-Identifier: BSD-2-Clause` instead. If there's a difference it's a judgement call if it could change the license or not.

jforissier commented:
> > And another, slightly related topic:
> > `plat-rcar/main.c` have full BSD-2 license at the top. If I get it correct, we can remove it only if all GlobaLogic agree on this. Or there is no such restriction?
> 
> If it matches word by word with the official BSD-2 license then it should be OK to use `SPDX-License-Identifier: BSD-2-Clause` instead. If there's a difference it's a judgement call if it could change the license or not.

Agreed, and in addition to that, it should preferably be a copyright owner, the one who added the full text in the first place, that removes the text. Which is the case here if I understand correctly. 
We have done this for Linaro files in the past.

jenswi-linaro commented on core/arch/arm/kernel/link_dummies_init.c:
```diff
@@ -6,7 +6,7 @@
 #include <mm/core_mmu.h>
 
-unsigned long __section(".text.dummy.get_aslr_seed")
+unsigned long __section(".text.dummy.get_aslr_seed") __weak
```
Functions in this file are used to override their normal counterpart as a way of breaking the dependency chain when partitioning the binary for paging. You need to find another way of overriding the `get_aslr_seed()` function.

jenswi-linaro reviewed (commented):
Comment for the commit "link_dummies: make get_aslr_seed() weak"

jenswi-linaro reviewed (commented):
Comment on commit message in "drivers: scif: update base address in scif_uart_init()"

Please make this message more clear by explaining what's wrong and what's fixed.

Prior to this patch...
So fix this by...

jenswi-linaro commented on core/arch/arm/plat-rcar/main.c:
```diff
@@ -49,6 +60,118 @@ register_dynamic_shm(NSEC_DDR_3_BASE, NSEC_DDR_3_SIZE);
+enum rcar_product_type rcar_get_product_type(void)
+{
+	static enum rcar_product_type rcar_product_type __nex_bss;
+	uint32_t prr;
```
Please initialize these two.

jenswi-linaro commented on core/arch/arm/plat-rcar/main.c:
```diff
@@ -49,6 +60,118 @@ register_dynamic_shm(NSEC_DDR_3_BASE, NSEC_DDR_3_SIZE);
+		switch (prr & PRR_CUT_MASK) {
+		case PRR_CUT_10:
+		{
+			uint32_t dummy5;
```
Please initialize these two.

jenswi-linaro commented on core/arch/arm/plat-rcar/main.c:
```diff
@@ -49,6 +60,118 @@ register_dynamic_shm(NSEC_DDR_3_BASE, NSEC_DDR_3_SIZE);
 static struct scif_uart_data console_data __nex_bss;
 
+enum rcar_product_type rcar_get_product_type(void)
+{
```
The implementation of this function is a bit verbose, but I can't see any way to significantly reduce it.

jenswi-linaro reviewed (commented):
Comments for "plat: rcar: add ability to determine SoC type in runtime"

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,124 @@
+				 uint64_t arg2, uint64_t arg3)
+{
+	/*
+	 * If MMU is enabled, we need to use trampoline function that will
```
Wouldn't it be nice to flatmap the needed parts of the ROM?
That place to do that would probably be in `core_init_mmu_map()`. I'm sure we can do something that isn't too ugly.

lorc replied:
It was my first intention.
We need to add at least two new memory types: 
 - identity-mapped MMIO ranges (because ROM code accesses MMIO using absolute addresses)
 - identity-mapped non-TEE RX memory (for ROM code itself)

And this would require some rework of `core_mmu.c`. If you think that this changes are okay, then I'll rewrite this PR.

jenswi-linaro replied:
With the MMIO ranges also, it might be a bit too much. My main concern is that we get too many ranges identity mapped. That could lead to `init_mem_map()` failing to find a suitable range if we're unlucky. Or leave us without a suitable location for the user mode table.

Another option could be to reimplement the hwrng driver in OP-TEE, but I guess you have your reasons to avoid that too. I'm fine with it as it is then.

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,124 @@
+	assert(rndbuff_pa % 8 == 0);
+	assert(scratch_pa % 8 == 0);
+
+	cache_op_inner(DCACHE_AREA_CLEAN, rndbuff, PLAT_RND_VECTOR_SZ);
```
I'm not sure this is needed for ARMv8.

lorc replied:
It appears, that it is need. Without this (but with cache invalidation afterwards) I'm getting stale data  in the buffer:

```
D/TC:0   test_hw_rng:1241 0000000056e32f70  b2 09 40 77 aa 5e 51 0c  03 dc 16 02 27 d5 9f 18                                                                                                   
D/TC:0   test_hw_rng:1241 0000000056e32f80  51 6c 1a ce 10 21 98 c5  ef ed a7 79 12 cc db bb 
D/TC:0   test_hw_rng:1241 0000000056e32f90  b2 09 40 77 aa 5e 51 0c  9f b4 b8 52 2f f0 ad fb 
D/TC:0   test_hw_rng:1241 0000000056e32fa0  0b 43 35 25 a0 4c 7a 0e  15 ac 02 89 0d d6 b7 a7 
D/TC:0   test_hw_rng:1241 0000000056e32fb0  5a c4 f8 66 68 b4 a4 74  6d e9 bc 45 92 56 ba 5e 
D/TC:0   test_hw_rng:1241 0000000056e32fc0  88 76 74 06 d5 e7 09 29  87 66 41 c1 ea 21 fe 8e 
D/TC:0   test_hw_rng:1241 0000000056e32fd0  1b 17 8a 3b 84 fa d5 3a  0b 9e e7 c7 fe a6 c2 49 
...
```

Note `b2 09 40 77...` sequence that appears twice

jenswi-linaro reviewed (commented):
Comments for "plat: rcar: add support for ROM API calls including HW RNG"

lorc commented:
I'm going to rework this PR, while I'll have free time

lorc commented:
It is still in progress. I rebased it on the master, but I need to address comments.

lorc commented:
Okay, I reworked the PR. I'm not targeting 3.13 so there is no rush in reviewing it.

Notable changes:

1. Added `plat_get_aslr_seed()` function
2. Removed that verbose function that determines SoC type and revision. 
3. Discovered and fixed issue with stack being corrupted when MMU is being turned off. Added comments describing why we need cache operations.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -103,6 +103,14 @@ __weak void main_secondary_init_gic(void)
 }
 
+/* May be overridden in plat-$(PLATFORM)/main.c */
+__weak unsigned long plat_get_aslr_seed(void)
```
Perhaps we could just rename `get_aslr_seed()` into this. That there's a `fdt` argument that might or might not mean anything shouldn't be a problem in platform specific code.

lorc replied:
I previously tried to provide own `get_aslr_seed()` in platform code.
But this required to mark `get_aslr_seed()` as a `__weak` function in `link_dummies_init.c` to resolve link issues. But you told that I can't do this. I'm talking about this comment: https://github.com/OP-TEE/optee_os/pull/4354#discussion_r565231088

Thus, I added that additional function `plat_get_aslr_seed`

lorc replied:
I previously tried to provide own `get_aslr_seed()` in platform code.
But this required to mark `get_aslr_seed()` as a `__weak` function in `link_dummies_init.c` to resolve link issues. But you told that I can't do this. I'm talking about this comment: https://github.com/OP-TEE/optee_os/pull/4354#discussion_r565231088

Thus, I added that additional function `plat_get_aslr_seed`

jenswi-linaro replied:
Oh, sorry. Find then.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1358,27 +1366,31 @@ unsigned long __weak get_aslr_seed(void *fdt)
 	if (rc) {
 		DMSG("Bad fdt: %d", rc);
-		return 0;
+		goto err;
```
I don't think it's meaningful with a fallback. Either are we're picking up the seed from DT or we have our own way, not point in both.

lorc replied:
What if OP-TEE is built with FDT support, but some platform revision does not provide it? This is sort of my case: some builds of ARM TF provide FDT, some - are not. Probably, there can be similar platforms.

It is debatable what should have higher priority - seed from FDT or `plat_` - provided one.

jenswi-linaro replied:
OK, if it's useful.

jenswi-linaro reviewed (commented):
Comments for "boot: introduce plat_get_aslr_seed()"

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,140 @@
+
+#include "romapi.h"
+
+extern uint32_t rcar_prr_value;
```
I would had preferred declaring this in a `.h` file.

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,140 @@
+	 * ensure that all writes reached memory. After the call we need to
+	 * invalidate the cache to see new data.
+	 *
+	 * We are not accessing scratch area, so no need to do cache maintenance
```
Then I believe the scratch buffer should be cache line aligned.

lorc replied:
Hm, not sure I get this.

This buffer never is being read or written by OP-TEE. ROM code needs it to store some own temporary data. So, why it should be cache line aligned?

jenswi-linaro replied:
What if there's a partial overlap with a cache line and some cached memory? If so you may have sudden corruptions in case that cache line is dirty and written into memory.

lorc replied:
Ah, I see. I had the same exact problem with a stack. So, `__plat_call_romapi` clears all data cache before jumping to ROM code. I do understand that this is sub-optimal, but I didn't found better solution.

jenswi-linaro replied:
If you make the memory areas cache line aligned too it should be water tight at least.
That stuff works is always a priority. ;-)

jenswi-linaro reviewed (commented):
Comments for "plat: rcar: add support for ROM API calls including HW RNG"

lorc commented:
Addressed the comments and also added new commit "plat: rcar: disable aarch32 support for OP-TEE core": ARM32 is not used by anyone and I don't want to write second `romapi_calls.S` for arm32. I will not be able to test it anyways.

lorc commented:
Hm, my local checkpatch does not produce the following error:

```
ERROR: space prohibited after that '&' (ctx:WxW)
#418: FILE: core/arch/arm/plat-rcar/romapi.c:23:
+	switch (rcar_prr_value & PRR_PRODUCT_MASK) {
 	                       ^
```

And  I'm not sure what is wrong with this piece of code. `checkpatch` assumes that I'm taking pointer to `PRR_PRODUCT_MASK` ?

jenswi-linaro commented:
Looks like a strange error to me, we're using that pattern at other places.

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,139 @@
+	paddr_t scratch_pa = va2pa(scratch);
+
+	assert(scratch_sz >= 4096);
+	assert(rndbuff_pa % 8 == 0);
```
Check that the buffers are cache line aligned instead?

lorc replied:
Well, I'm clearing the whole dcache as you saw. Or `cache_op_inner()` have requirement for cache alignment?

jenswi-linaro replied:
What if there's a partial overlap so there's a shared cache line with something used by another core?

lorc replied:
Ugh, caches are really hard :)

So, you want to say that  with caches disabled, our core is no more part of cacheability domain. And some other core can read-modify-write adjacent memory, rewriting part of our buffer with old values. This is interesting case, yes.

I'll enforce alignment check. Thanks for the enlightenment :)

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,139 @@
+	assert(rndbuff_pa % 8 == 0);
+	assert(scratch_pa % 8 == 0);
+
+	ret = plat_call_romapi(func_addr, rndbuff_pa, scratch_pa, scratch_sz);
```
I believe we need to invalidate `rndbuf`before calling this function too, in case the buffer has dirty cache lines.
It is important the `rndbuf` is cache line aligned since we may otherwise affect unrelated memory.

lorc replied:
Yep, I'm clearing the whole dcache. Otherwise I experienced dirty cache writes to a  stack.

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi_call.S:
```diff
@@ -0,0 +1,87 @@
+
+	/* Disable MMU */
+	mrs	x9, sctlr_el1
+	and	x9, x9, #(~SCTLR_M)
```
`bic` is otherwise the usual instruction for this.

lorc replied:
Good idea. Will change.

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi_call.S:
```diff
@@ -0,0 +1,87 @@
+	/* Disable MMU */
+	mrs	x9, sctlr_el1
+	and	x9, x9, #(~SCTLR_M)
+	and	x9, x9, #(~SCTLR_C)
```
Why do you need to turn off the data cache? Some requirement by the ROM? If you were able to keep it on there's a good chance that you wouldn't need to do any cache maintenance at all.

lorc replied:
Hmmm, maybe my understanding of VMSA is incorrect. When MMU is disabled, all memory is treated as Device-nGnRnE, so no data cache can be used. I'm not sure how data cache will behave in this case. 

According to D4.10.3 "If the change affects the cacheability attributes of the area of memory, including any change between Write-Through and Write-Back attributes, software must ensure that any cached copies of affected locations are removed from the caches, typically by cleaning and invalidating the locations from the levels of cache that might hold copies of the locations affected by the attribute change". I'm assuming that disabling MMU can be considered as "changing cacheability attributes".

So, maybe it is not mandatory to clear C bit. But I need to do perform cache maintenance anyways. If my understanding of ARM ARM is correct, of course :)

jenswi-linaro replied:
I believe this changed with Aarch64, but I don't have a reference right now.

lorc replied:
Well, practice is the criterion of truth :) OP-TEE panics if I'm not clearing `SCTLR_C`

But, now I'm seeing a new issue which I didn't saw earlier. I need to investigate it...

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi_call.S:
```diff
@@ -0,0 +1,87 @@
+	isb
+
+	/* Invalidate instruction cache and branch predictor */
+	ic	iallu
```
`ialluis` or `iallu`?
Isn't the `dsb ish` or `dsb` missing?

lorc replied:
In my understanding, I didn't touched page tables, so `dsb` is not required. But I may be wrong there. MMU is hard :)

jenswi-linaro commented:
For the commits
"drivers: scif: update base address in scif_uart_init()"
"plat: rcar: disable aarch32 support for OP-TEE core"
"plat: rcar: store PRR value in global variable"
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

For the commit "boot: introduce plat_get_aslr_seed()"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
It is strange indeed, note that the error doesn't show up when checking the individual commits but only when checking the whole diff. It looks like `checkpatch` is confused by a previous patch. In other words patch P1 is OK, patch P2 has two unrelated warnings (not the `&` thing), and patch P1 + P2 show this `&` warning. I could reproduce the issue locally:

```
# Checking P1
$ git diff a8657757^^...a8657757^ | ../linux/scripts/checkpatch.pl 
$

# Checking P2
$ git diff a8657757^...a8657757 | ../linux/scripts/checkpatch.pl 
WARNING: else is not generally useful after a break or return
#139: FILE: core/arch/arm/plat-rcar/romapi.c:80:
+		return fptr(func, arg1, arg2, arg3);
+	} else {

WARNING: 'becase' may be misspelled - perhaps 'because'?
#269: FILE: core/arch/arm/plat-rcar/romapi_call.S:33:
+	 * becase assemlber will generate relative address to virt_to_phys(),

total: 0 errors, 2 warnings, 0 checks, 294 lines checked

# Checking P1 + P2
$ git diff a8657757^^...a8657757 | ../linux/scripts/checkpatch.pl 
ERROR: space prohibited after that '&' (ctx:WxW)
#136: FILE: core/arch/arm/plat-rcar/romapi.c:23:
+	switch (rcar_prr_value & PRR_PRODUCT_MASK) {
 	                       ^

ERROR: space prohibited after that '&' (ctx:WxW)
#138: FILE: core/arch/arm/plat-rcar/romapi.c:25:
+		switch (rcar_prr_value & PRR_CUT_MASK) {
 		                       ^

ERROR: space prohibited after that '&' (ctx:WxW)
#152: FILE: core/arch/arm/plat-rcar/romapi.c:39:
+		switch (rcar_prr_value & PRR_CUT_MASK) {
 		                       ^

WARNING: else is not generally useful after a break or return
#193: FILE: core/arch/arm/plat-rcar/romapi.c:80:
+		return fptr(func, arg1, arg2, arg3);
+	} else {

WARNING: 'becase' may be misspelled - perhaps 'because'?
#323: FILE: core/arch/arm/plat-rcar/romapi_call.S:33:
+	 * becase assemlber will generate relative address to virt_to_phys(),

total: 3 errors, 2 warnings, 0 checks, 331 lines checked
```

So let's not worry about this.

Now I need to understand why my local `checkpatch` found a spelling mistake but Travis did not...

jforissier commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,139 @@
+		 * at identity address, otherwise it will experience problem
+		 * when MMU will be disabled.
+		 */
+		fptr = (void *)virt_to_phys(__plat_call_romapi);
```
Why is `virt_to_phys()` needed here?

lorc replied:
After relocation (if ASLR is enabled), `__plat_call_romapi` function will be mapped at two places: relocated and identity-mapped. But `__plat_call_romapi` symbol will point to relocated address. Correct me if I'm wrong.

I need to call this function by identity mapped address, because it will disable MMU for time being. `virt_to_phys` returns physical address, which is the same as virtual for identity-mapped memory.

jforissier replied:
Ah, right, now that I read the comment again it makes sense. When I read this initially, for some reason I did not realize the function has two virtual addresses. I focused on the identity mapping and therefore the `virt_to_phys()` call looked like a no-op to me. I missed the fact that the symbol has the relocated value here. Perhaps add this to the comment for people like me...?

lorc replied:
Sure :)

I elaborated the comment. Also I addressed @jenswi-linaro's comments regarding cache line alignments and fixed a new issue with call to `dcache_op_all` from identity location.

jforissier commented:
- For "drivers: scif: update base address in scif_uart_init()",
      "boot: introduce plat_get_aslr_seed()",
      "plat: rcar: implement hw_get_random_byte() function":
  `Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For "plat: rcar: disable aarch32 support for OP-TEE core",
      "plat: rcar: store PRR value in global variable":
  `Acked-by: Jerome Forissier <jerome@forissier.org>`

lorc commented:
Okay, so I squashed fixups and added  tags for already reviewed patches

jforissier commented:
"plat: rcar: enable ASLR on the platform" is:

`Acked-by: Jerome Forissier <jerome@forissier.org>`

(you might want to strip "on the platform" from the subject)

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,144 @@
+
+		return fptr(func, arg1, arg2, arg3);
+	} else {
+		uint32_t (*fptr)(uint64_t arg1, uint64_t arg2, uint64_t arg3);
```
I didn't notice this before, but it seems like the supplied function pointer `func` is called using different argument depending on if mmu is enabled of not.

lorc replied:
Yes. If MMU is disabled, we are jumping straight to a ROM code there. If MMU is enabled, we using the trampoline is `.S` file. Trampoline needs ROM code address, so there is one extra parameter.

jenswi-linaro reviewed (commented):
Comment for the commit "plat: rcar: add support for ROM API calls including HW RNG"

jforissier commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -0,0 +1,144 @@
+	} else {
+		uint32_t (*fptr)(uint64_t arg1, uint64_t arg2, uint64_t arg3);
+
+		fptr = (typeof(fptr))func;
```
In the first branch you're casting to `void *`, I think both should be the same.
In fact I would recommend addressing the checkpatch warning, the code would be more clear with two variables of different names:
```C
        uint32_t (*tramp)(paddr_t func, uint64_t arg1, uint64_t arg2, uint64_t arg3) = NULL;
        uint32_t (*fptr)(uint64_t arg1, uint64_t arg2, uint64_t arg3) = (void *)func;

        if (cpu_mmu_enabled()) {
                tramp = ...
                return tramp(func, arg1, arg2, arg3);
        }
        return fptr(arg1, arg2, arg3);

lorc replied:
I reworked it slightly different way: added a second trampoline function with the same parameters list. So now I have one function pointer for both variants. 

I pushed this change as fixup.

jforissier replied:
Fine with me.

jforissier commented on core/arch/arm/plat-rcar/romapi.c:
```diff
@@ -54,41 +55,51 @@ static int get_api_table_index(void)
+	if (cpu_mmu_enabled())
 		/*
 		 * With ASLR enabled __plat_call_romapi() function will be
 		 * mapped at two addresses: at random address (with the rest of
```
__plat_romapi_wrapper

jforissier commented:
Could you remove the unnecessary backslashes in commit "plat: rcar: implement hw_get_random_byte() function"? (checkpatch).

For commit "plat: rcar: add support for ROM API calls including HW RNG":
`Acked-by: Jerome Forissier <jerome@forissier.org>`

lorc commented:
> Could you remove the unnecessary backslashes in commit "plat: rcar: implement hw_get_random_byte() function"? (checkpatch).

Well, I removed them and now checkpatch is not happy about line length :)

jenswi-linaro commented:
For the commit "plat: rcar: add support for ROM API calls including HW RNG" please apply:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

lorc commented:
Applied the tag and returned backslashes in "plat: rcar: implement hw_get_random_byte() function"

jforissier commented:
> Applied the tag and returned backslashes in "plat: rcar: implement hw_get_random_byte() function"

Why the backslash ? It is not a macro...

lorc commented:
> Why the backslash ? It is not a macro...

Because of unexpected spike of stupidity from my side :) 

Thanks for the tip. I fixed it now.

jforissier commented:
> > Why the backslash ? It is not a macro...
> 
> Because of unexpected spike of stupidity from my side :)
> 
> Thanks for the tip. I fixed it now.

No worries :)
